[CodeGen] Allow negative frame indicies in Register class.

mgudim opened the pull request:
The register values between `2 << 30` (inclusive) and `2 << 31` (exclusive) correspond to frame indices. To obtain the frame index from the given register value we interpret  first 30 bits as an unsigned integer. Thus, currently only non-negative frame indices can be represented.  

However, we should also be able to represent negative frame indices as register values as well. This is used by reaching definitions analysis for example.

In order to do that, we interpret the first 30 bits of the register value as a signed integer.

Commit:
Negative frame indicies as register.

This is used by reaching definitions analysis in order to track stores /
loads from negative frame indices.

mgudim commented:
this is part of work for: https://github.com/llvm/llvm-project/pull/90819
but hopefully it is useful on its own.

topperc commented on llvm/include/llvm/CodeGen/Register.h:
```diff
@@ -87,7 +91,7 @@ class Register {
   int stackSlotIndex() const {
     assert(isStack() && "Not a stack slot");
-    return static_cast<int>(Reg - Register::FirstStackSlot);
+    return static_cast<int>(SignExtend64(Reg & Register::StackSlotMask, 30));
```
Use `SignExtend64<30>`

Commit:
addressed review comments.

arsenm commented on llvm/include/llvm/CodeGen/Register.h:
```diff
@@ -87,7 +91,7 @@ class Register {
   int stackSlotIndex() const {
     assert(isStack() && "Not a stack slot");
-    return static_cast<int>(Reg - Register::FirstStackSlot);
+    return static_cast<int>(SignExtend64<30>(Reg & Register::StackSlotMask));
```
```suggestion
    return static_cast<int>(SignExtend64<MaxFrameIndexBitwidth>(Reg & Register::StackSlotMask));
```

arsenm reviewed (commented):
I don't understand the title. Can you rephrase an elaborate in the description

Commit:
Renamed "FirstStackSlot" to "StackSlotZero".
Simplified the definition of "StackSlotMask"
Fixed wrong assertion.
Added unit test.

topperc commented on llvm/include/llvm/CodeGen/Register.h:
```diff
@@ -87,7 +92,8 @@ class Register {
     assert(isStack() && "Not a stack slot");
-    return static_cast<int>(Reg - Register::FirstStackSlot);
+    return static_cast<int>(
+        SignExtend64<MaxFrameIndexBitwidth>(Reg & Register::StackSlotMask));
```
Use SignExtend32 so you don't need the static_cast?

Commit:
addressed review comments.

Commit:
address review comments.

topperc commented on llvm/include/llvm/CodeGen/Register.h:
```diff
@@ -35,19 +36,23 @@ class Register {
-    assert(FI >= 0 && "Cannot hold a negative frame index.");
-    return Register(FI + Register::FirstStackSlot);
+    unsigned FIMasked = FI & Register::StackSlotMask;
+    assert(isUInt<MaxFrameIndexBitwidth>(FIMasked) &&
```
I think we need to check isInt<MaxFrameIndexBitwidth> before masking it. Doesn't the masking guarantee the isUInt check passes?

topperc commented:
> I don't understand what a "negative frame index" is supposed to be. The `Register` class can encode references to stack slots which are indices into an array of stack objects in `MachineFrameInfo` and negative indexes make no sense there, no?

From MachineFrameInfo.h

```
/// To support this, the class assigns unique integer identifiers to stack       
/// objects requested clients.  These identifiers are negative integers for      
/// fixed stack objects (such as arguments passed on the stack) or nonnegative   
/// for objects that may be reordered.  Instructions which refer to stack        
/// objects use a special MO_FrameIndex operand to represent these frame         
/// indexes.  
```

arsenm commented on llvm/unittests/CodeGen/RegisterTest.cpp:
```diff
@@ -0,0 +1,33 @@
+
+  for (int64_t FI : FIs) {
+    Reg = Register::index2StackSlot(FI);
+    ASSERT_EQ(Reg.stackSlotIndex(), FI);
```
```suggestion
    EXPECT_EQ(Reg.stackSlotIndex(), FI);
```

Commit:
addressed review comments.

topperc commented:
> > > I don't understand what a "negative frame index" is supposed to be. The `Register` class can encode references to stack slots which are indices into an array of stack objects in `MachineFrameInfo` and negative indexes make no sense there, no?
> > 
> > 
> > From MachineFrameInfo.h
> > ```
> > /// To support this, the class assigns unique integer identifiers to stack       
> > /// objects requested clients.  These identifiers are negative integers for      
> > /// fixed stack objects (such as arguments passed on the stack) or nonnegative   
> > /// for objects that may be reordered.  Instructions which refer to stack        
> > /// objects use a special MO_FrameIndex operand to represent these frame         
> > /// indexes.  
> > ```
> 
> @MatzeB @topperc @arsenm We expect vast majority of the indices to be positive and a few negative ones. In theory we could "shift" the values so that we use the same 30 bits but only represent say 100 negative values and the rest are positive. I can do this too if people think its necessary.

I don't think its necessary. Having more than 2^29 frame indices in either direction seems unlikely. The amount of memory consumed by MachineFrameInfo just to track that would be excessive.

topperc commented on llvm/unittests/CodeGen/RegisterTest.cpp:
```diff
@@ -0,0 +1,34 @@
+}
+
+TEST(RegisterTest, StackSlotIndex) {
+  std::vector<int64_t> FIs = {0, 1 - 1, (1 << 29) - 1, -(1 << 29)};
```
Can we use Register::MaxFrameIndexBitwidth-1 instead of 29?

Commit:
address review comments.

topperc commented on llvm/unittests/CodeGen/RegisterTest.cpp:
```diff
@@ -0,0 +1,35 @@
+  // check that we do not crash on the highest possible value of frame index.
+  EXPECT_NO_FATAL_FAILURE(Register::index2StackSlot((1 << 29) - 1));
+  // check that we do not crash on the lowest possible value of frame index.
+  EXPECT_NO_FATAL_FAILURE(Register::index2StackSlot(-(1 << 29)));
```
It should use `Register::MaxFrameIndexBitwidth - 1` instead of putting 29 anywhere in the test.

ppenzin replied:
I've pushed an edit to this and the other line.

ppenzin committed:
Review feedback: more MaxFrameIndexBitwidth - 1

arsenm commented on llvm/unittests/CodeGen/RegisterTest.cpp:
```diff
@@ -0,0 +1,36 @@
+  int MaxPowOf2 = 1 << (Register::MaxFrameIndexBitwidth - 1);
+  // check that we do not crash on the highest possible value of frame index.
+  EXPECT_NO_FATAL_FAILURE(Register::index2StackSlot(MaxPowOf2 - 1));
+  // check that we do not crash on the lowest possible value of frame index.
```
```suggestion
  // Check that we do not crash on the lowest possible value of frame index.
```

arsenm commented on llvm/unittests/CodeGen/RegisterTest.cpp:
```diff
@@ -0,0 +1,36 @@
+            Register::StackSlotZero | Register::StackSlotMask);
+  int MaxPowOf2 = 1 << (Register::MaxFrameIndexBitwidth - 1);
+  // check that we do not crash on the highest possible value of frame index.
+  EXPECT_NO_FATAL_FAILURE(Register::index2StackSlot(MaxPowOf2 - 1));
```
Why EXPECT_NO_FATAL_FAILURE instead of just checking the actual value?

ppenzin committed:
Review feedback - min and max stack values

topperc reviewed (approved):
LGTM

ppenzin committed:
Merge branch 'main' into neg_indices

ppenzin committed:
Merge branch 'main' into neg_indices
